frontend/src/metabase/common/components/EntityPicker/types.ts

```diff
@@ -24,11 +25,31 @@ export type PickerStateItem<Item, Query> = {
 
 export type EntityPickerOptions = EntityPickerModalOptions;
 
-export type EntityTab<Model extends string> = {
+export type EntityPickerTabRenderProps<
+  Id extends SearchResultId,
+  Model extends string,
+  Item extends TypeWithModel<Id, Model>,
+> = {
+  onItemSelect: (item: Item) => void;
+};
+
+export type EntityPickerTabId = string;
+
+export type EntityPickerTab<
+  Id extends SearchResultId,
+  Model extends string,
+  Item extends TypeWithModel<Id, Model>,
+> = {
+  id: EntityPickerTabId;
   displayName: string;
-  element: JSX.Element;
+  render: (props: EntityPickerTabRenderProps<Id, Model, Item>) => JSX.Element;
```

Comment:
I really wanted to keep the new `tabFolderState` inside `EntityPickerModal`. There's no reason for this state to live outside of `EntityPickerModal`.

The problem was that the tab components are always defined outside of `EntityPickerModal` (well, except recents and search tabs), so `EntityPickerModal` was not aware of what was happening inside those tabs (like: folder being clicked). 

With this render pattern, `EntityPickerModal` can now inject its `onItemSelect` and monitor what's being clicked inside the tabs.

----

There's an interesting consequence of this change. Previously you had to pass your `onItemSelect` handler to `EntityPickerModal` and every single tab component. Now you only need to pass it to `EntityPickerModal`.